_drafts/printf-on-embedded.md

```diff
@@ -0,0 +1,566 @@
+won't stomp on the global reentrancy data if it preempted an in-progress
+non-reentrant function!
+
+Note that you'll need to do the same thing in every preemption context, i.e. any
+call site that can interrupt another thread that is using non-reentrant
+functions. So this approach would be primarily useful in cases with only a small
+number of threads, or a well-considered preemption hierarchy where we know
+exactly where to move the pointer and where we don't need to.
+
+## Redirecting stdio
+
+This section describes a few approaches for redirecting libc standard
+input/output functions, and compares the tradeoffs each one makes.
+
+| Method      | Advantages                                                                              | Disadvantages                                                                              |
+| ----------- | --------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------ |
+| UART        | - simple, no special host HW needed                                                     | - requires spare UART + pins on target<br/>- can be slow                                   |
+| Semihosting | - no extra HW requirements on target                                                    | - requires a debug probe for the host to use<br/>- _prohibitively_ slow, interrupts target |
+| SWO         | - requires only a single (often specific) spare pin on the target<br/>- relatively fast | - usually needs a debug probe for the host to use<br/>- output only                        |
+| RTT         | - no extra HW requirements on target<br>- relatively fast                               | - requires a debug probe for the host to use<br/>- license may be problematic              |
+
+### UART
+
+UART: asynchronous serial. Most embedded microcontrollers will have at least
+one.
+
+Pros:
+
+- simple and widely available protocol (lots of available software and hardware
+  tools interface to it)
+- doesn't require an attached debugger; you can use it in PROD 😀
+
+Cons:
+
+- requires extra hardware to interface to the PC host, typically (USB to
+  asynchronous serial adapter, like the FTDI or CP2102 adapters)
+- requires configuring and using the UART peripheral (if your microcontroller
+  doesn't have many UARTs, might be a problem using it for `printf`)
+
+The biggest downside is you're spending one of your UART peripherals for printf,
+which might be needed for your actual application.
+
+Also, configuring a UART might require managing clock configuration, chip
+power/sleep modes, figuring out buffering the data (DMA?), making sure
+everything is interrupt safe, etc.
+
+And finally, you'll usually need some kind of adapter dongle to interface the
+UART with your PC (eg USB-to-TTL or USB-to-asynchronous serial).
+
+But! because the interface is contained entirely on the microcontroller, it can
```

Comment:
Exclamation point is kind of weird, but I won't protest if you want to leave it for flavor